src/control/server/storage/bdev/backend_class.go

```diff
@@ -116,36 +112,82 @@ func writeConf(log logging.Logger, templ string, req *FormatRequest) error {
 		req.OwnerUID, req.OwnerGID)
 }
 
-// writeNvmeConf generates nvme config file for given bdev type to be consumed
-// by spdk.
-func (sb *spdkBackend) writeNvmeConfig(req *FormatRequest) error {
-	if req.ConfigPath == "" {
-		return errors.New("no output config directory set in request")
-	}
-
+func writeIniConfig(log logging.Logger, enableVmd bool, req FormatRequest) error {
 	templ := map[storage.BdevClass]string{
 		storage.BdevClassNvme: clsNvmeTemplate,
 		storage.BdevClassKdev: clsKdevTemplate,
 		storage.BdevClassFile: clsFileTemplate,
 	}[req.Class]
 
 	// special handling for class nvme
-	if req.Class == storage.BdevClassNvme {
-		if len(req.DeviceList) == 0 {
-			sb.log.Debug("skip write nvme conf for empty device list")
-			return nil
-		}
-		if !sb.IsVMDDisabled() {
-			templ = `[Vmd]
+	if req.Class == storage.BdevClassNvme && enableVmd {
+		templ = `[Vmd]
     Enable True
 
 ` + templ
-		}
 	}
 
 	// spdk ini file expects device size in MBs
 	req.DeviceFileSize = req.DeviceFileSize / humanize.MiByte
 
-	sb.log.Debugf("write nvme output config: %+v", req)
-	return writeConf(sb.log, templ, req)
+	buf, err := renderTemplate(&req, templ)
+	if err != nil {
+		return err
+	}
+	if err := writeConfFile(log, &buf, &req); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func writeJsonConfig(log logging.Logger, enableVmd bool, req FormatRequest) error {
+	nsc := newNvmeSpdkConfig(req.DeviceList, req.Hostname)
+
+	buf, err := json.MarshalIndent(nsc, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	if err := writeConfFile(log, bytes.NewBuffer(buf), &req); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// writeNvmeConf generates nvme config file for given bdev type to be consumed
+// by spdk.
+func (sb *spdkBackend) writeNvmeConfig(req *FormatRequest) error {
+	if req.ConfigPath == "" {
+		return errors.New("no output config directory set in request")
+	}
+	if req.Class == storage.BdevClassNvme && len(req.DeviceList) == 0 {
+		sb.log.Debug("skip write nvme conf for empty device list")
+		return nil
+	}
+
+	enableVmd := !sb.IsVMDDisabled()
+
+	sb.log.Debugf("write nvme output ini config: %+v", req)
+	// pass request by value to restrict the scope of side effects
+	if err := writeIniConfig(sb.log, enableVmd, *req); err != nil {
+		return err
+	}
```

Comment:
I'm a bit confused by this. Why are you writing the ini config and potentially also the json config?

Reply:
I want to land the upgrade in an iterative fashion, landing this will enable compatibility with both bio versions , I don't want to land all the work at the same time which will  probably only all be ready so close to the release deadline.  I want to leave time for each of the individual parts to soak for as long as possible to avoid destabilisation.